### PR TITLE
fix incorrect types generated when refs reference types with discriminators

### PR DESCRIPTION
**Pull request by adamschoenemann**
## Changes

This PR fixes the code, so that incorrect types are not generated when referencing a type (with a discriminator with an implicit mapping) inside a oneOf/anyOf

## How to Review

I added tests to both `index.test.ts` and  `schema-object.test.ts`, and I tried to fix the code, but I must admit it was kind of a shot in the dark, but it does seem to work :)

## Checklist

- [x] Unit tests updated
- [x] ~~`docs/` updated (if necessary)~~
- [x] `pnpm run update:examples` run (only applicable for openapi-typescript) (I ran this but nothing changed)

### CODE REVIEW

**Comment by adamschoenemann** [comment]
Test macos failed, because the digital ocean test timed out. I'm pretty sure it's got nothing to do with this change... I observed this test being quite slow locally as well, on the main branch

File: packages/openapi-typescript/test/index.test.ts [hunk 1]
```diff
@@ -625,6 +625,93 @@ export interface components {
 
 export type external = Record<string, never>;
 
+export type operations = Record<string, never>;
+`);
+    });
+
+    test("discriminator with implicit mapping (oneOf)", async () => {
```
Comment (drwpow) [inline packages/openapi-typescript/test/index.test.ts:632]: Awesome tests!

File: packages/openapi-typescript/test/schema-object.test.ts [hunk 2]
```diff
@@ -374,48 +374,6 @@ describe("Schema Object", () => {
         expect(generated).toBe("0 | 1");
       });
 
-      test("empty object + oneOf is ignored", () => {
```
Comment (drwpow) [inline packages/openapi-typescript/test/schema-object.test.ts:377]: We’ll want to keep this test—this is intended behavior.

Comment (adamschoenemann) [reply packages/openapi-typescript/test/schema-object.test.ts:377]: Oh, I didn't mean to delete this test at all! I'll fix it when I'm back at my laptop tomorrow.
Also, thanks for the kind words, much appreciated ☺

File: packages/openapi-typescript/test/index.test.ts (hunk 1 above)
Comment (adamschoenemann) [reply packages/openapi-typescript/test/index.test.ts:632]: Thanks!

File: packages/openapi-typescript/test/schema-object.test.ts (hunk 2 above)
Comment (adamschoenemann) [reply packages/openapi-typescript/test/schema-object.test.ts]: Un-deleted the test now :)

**Review by drwpow — APPROVED** [review]
Awesome work! 👏 Thank you

**Comment by drwpow** [comment]
Just a heads up: In reviewing issues I realized #1158 does outline a bug that results in slightly-altered typegen in one of the tests added:

```diff
- components["schemas"]["parent"] | null
+ {
  operation: "schema-object";
} & (components["schemas"]["parent"] | null)
```

This seems weird, but I believe the discriminator spec does suggest that **if `propertyName` is present, it should generate that property even with no mapping**.

It’s weird, but seems to be how it’s outlined 🤷. Anyways that shouldn’t cause a regression in this fix at all; just wanted to flag an incoming change.

**Pull request merged**